### ISSUE TITLE
Error Domain=com.theo.network.error Code=-1011

### ISSUE THREAD

**Issue by nskostas** [GraphStory/neo4j-ios#12]
Hi, I'm new with neo4j and GraphStory, I'm going through the steps in the readme file, and I'm getting the following error.

```
meta in success nil error Optional(Error Domain=com.theo.network.error Code=-1011 "The operation couldn’t be completed. (com.theo.network.error error -1011.)" UserInfo=0x7f93c1418380 {TheoResponse=<NSHTTPURLResponse: 0x7f93c15871b0> { URL: https://neo-559156611beb5-364459c455.do-stories.graphstory.com:7473//db/data/ } { status code: 401, headers {
    Connection = "keep-alive";
    "Content-Type" = "application/json; charset=UTF-8";
    Date = "Fri, 03 Jul 2015 16:12:46 GMT";
    Server = "nginx/1.7.6";
    "Transfer-Encoding" = Identity;
    "Www-Authenticate" = None;
} }, TheoResponseCode=401, NSLocalizedDescriptionKey=The response was empty, but you received at valid response code})
meta in success nil error Optional(Error Domain=com.theo.network.error Code=-1 "The operation couldn’t be completed. (com.theo.network.error error -1.)" UserInfo=0x7f93c141c5b0 {TheoResponse=<NSHTTPURLResponse: 0x7f93c15871b0> { URL: https://neo-559156611beb5-364459c455.do-stories.graphstory.com:7473//db/data/ } { status code: 401, headers {
    Connection = "keep-alive";
    "Content-Type" = "application/json; charset=UTF-8";
    Date = "Fri, 03 Jul 2015 16:12:46 GMT";
    Server = "nginx/1.7.6";
    "Transfer-Encoding" = Identity;
    "Www-Authenticate" = None;
} }, TheoResponseCode=401, NSLocalizedDescriptionKey=There was an error processing the request})
```

In viewDidLoad() I execute the following code : 
```
        let theo: Client = Client(baseURL: "https://neo-559156611beb5-364459c455.do-stories.graphstory.com:7473/", user: "559156611beb5", pass: "r3A8Ge1jx8KKzFmG1RLEZK2daHXWkZHlwTnK7J0G")
       
        theo.metaDescription({(meta, error) in
            println("meta in success \(meta) error \(error)")
        })
```

Any suggestions?

Xcode 6.4, iOS SDK 8.4, OSX 10.10.4

Thank you,
Kind regards,
Kostas.

**Comment by funkatron** [comment]
I'm going to add the info from your support emails here, and try to get others to participate. I think we solved your 401 error, but we hit another roadblock.

--------------------

I think you were correct, but now I’m receiving a nil response from the following request: 

```swift
let task : NSURLSessionDataTask = self.httpSession.session.dataTaskWithRequest(request, 
completionHandler: {(data: NSData!, response: NSURLResponse!, error: NSError!)
```

The value of the request is:

```
 <NSMutableURLRequest: 0x7f9d12430320> { URL: neo-559156611beb5-364459c455.do-stories.graphstory.com:7473/db/data/ }
```

Thank you very much for your time, 
sorry for the trouble.

Very much appreciate your help,
Kostas.

**Comment by kwylez** [comment]
@funkatron is the above value for the request from the `metaDescription` request?

**Comment by nskostas** [comment]
@kwylez yes, it's from ```metaDescription``` request, and then ```getResource```.

**Comment by kwylez** [comment]
@nskostas I believe I know what is going on. For some reason the credential storage isn't being used by the session...thus, the **401**. I'm not sure when this was introduced. There are references to related issues with iOS 8. With that said the official Apple docs state that the session configuration can't be modified with a "live" session which would explain what is happening.

I'm looking into a fix as we speak

**Comment by kwylez** [comment]
@nskostas what version of neo4j is your instance running?

**Comment by nskostas** [comment]
@kwylez  I don't know how to check the version in the Graphstory, but locally it was Neo4j CE 2.2.3 for sure.

**Comment by jeremykendall** [comment]
@nskostas You're on Enterprise 2.2.3

/cc @kwylez

**Comment by kwylez** [comment]
@nskostas After debugging this all afternoon it turns out that Neo4j 2.2.x is not properly handling the basic auth handshake thus the reason why you are getting that error. Instances running 2.1.x work great with Theo. I'm working on a fix for this and should have something ready to test by Monday.

**Comment by kwylez** [comment]
@nskostas I've added in the fixes and they are in the `issue/12` branch. please use that branch to verify it is working for you. Once I hear back I'll be able to merge everything into `master`

**Comment by nskostas** [comment]
@kwylez I've run the tests but it returns an ```Thread 1: EXC_BAD_INSTRUCTION```, in ```Line 152``` of the ```Request.swift```. This is the console output: 
``` Test Suite 'All tests' started at 2015-07-21 10:48:49 +0000
Test Suite 'TheoTests.xctest' started at 2015-07-21 10:48:49 +0000
Test Suite 'Theo_000_RequestTests' started at 2015-07-21 10:48:49 +0000
Test Case '-[TheoTests.Theo_000_RequestTests test_000_successfullyFetchDBMeta]' started.
fatal error: unexpectedly found nil while unwrapping an Optional value```

I've also tried a sample example, with a new ```theo: Client``` and a ```theo.metaDescription``` call, and it return an error: ```fatal error: unexpectedly found nil while unwrapping an Optional value``` in ```Line 102```, of ```Request.swift```.

Am I doing anything wrong, do you want me to try something different?

**Comment by kwylez** [comment]
Can you go to your terminal and run `git branch`

**Comment by nskostas** [comment]
```
$ git branch
  develop/swift-1.2
* issue/12
  master
```

**Comment by kwylez** [comment]
Are you running just the first test or all of them and the first test is the only one that fails? 

Just running the first test for me works just fine.

**Console output**

*neo4j_version: 2.2.3*
```
Test Suite 'Selected tests' started at 2015-07-21 15:48:46 +0000
Test Suite 'TheoTests.xctest' started at 2015-07-21 15:48:46 +0000
Test Suite 'Theo_000_RequestTests' started at 2015-07-21 15:48:46 +0000
Test Case '-[TheoTests.Theo_000_RequestTests test_000_successfullyFetchDBMeta]' started.
meta in success Optional(Extensions: [SpatialPlugin: {
    addCQLDynamicLayer = "https://neo-55a53f8980d73-364459c455.do-stories.graphstory.com:7473/db/data/ext/SpatialPlugin/graphdb/addCQLDynamicLayer";
    addEditableLayer = "https://neo-55a53f8980d73-364459c455.do-stories.graphstory.com:7473/db/data/ext/SpatialPlugin/graphdb/addEditableLayer";
    addGeometryWKTToLayer = "https://neo-55a53f8980d73-364459c455.do-stories.graphstory.com:7473/db/data/ext/SpatialPlugin/graphdb/addGeometryWKTToLayer";
    addNodeToLayer = "https://neo-55a53f8980d73-364459c455.do-stories.graphstory.com:7473/db/data/ext/SpatialPlugin/graphdb/addNodeToLayer";
    addNodesToLayer = "https://neo-55a53f8980d73-364459c455.do-stories.graphstory.com:7473/db/data/ext/SpatialPlugin/graphdb/addNodesToLayer";
    addSimplePointLayer = "https://neo-55a53f8980d73-364459c455.do-stories.graphstory.com:7473/db/data/ext/SpatialPlugin/graphdb/addSimplePointLayer";
    findClosestGeometries = "https://neo-55a53f8980d73-364459c455.do-stories.graphstory.com:7473/db/data/ext/SpatialPlugin/graphdb/findClosestGeometries";
    findGeometriesInBBox = "https://neo-55a53f8980d73-364459c455.do-stories.graphstory.com:7473/db/data/ext/SpatialPlugin/graphdb/findGeometriesInBBox";
    findGeometriesWithinDistance = "https://neo-55a53f8980d73-364459c455.do-stories.graphstory.com:7473/db/data/ext/SpatialPlugin/graphdb/findGeometriesWithinDistance";
    getLayer = "https://neo-55a53f8980d73-364459c455.do-stories.graphstory.com:7473/db/data/ext/SpatialPlugin/graphdb/getLayer";
    updateGeometryFromWKT = "https://neo-55a53f8980d73-364459c455.do-stories.graphstory.com:7473/db/data/ext/SpatialPlugin/graphdb/updateGeometryFromWKT";
}] node: https://neo-55a53f8980d73-364459c455.do-stories.graphstory.com:7473/db/data/node node_index: https://neo-55a53f8980d73-364459c455.do-stories.graphstory.com:7473/db/data/index/node relationship_index: https://neo-55a53f8980d73-364459c455.do-stories.graphstory.com:7473/db/data/index/relationship extensions_info : https://neo-55a53f8980d73-364459c455.do-stories.graphstory.com:7473/db/data/ext, relationship_types: https://neo-55a53f8980d73-364459c455.do-stories.graphstory.com:7473/db/data/relationship/types batch: https://neo-55a53f8980d73-364459c455.do-stories.graphstory.com:7473/db/data/batch cypher: https://neo-55a53f8980d73-364459c455.do-stories.graphstory.com:7473/db/data/cypher indexes: https://neo-55a53f8980d73-364459c455.do-stories.graphstory.com:7473/db/data/schema/index constraints: https://neo-55a53f8980d73-364459c455.do-stories.graphstory.com:7473/db/data/schema/constraint transaction: https://neo-55a53f8980d73-364459c455.do-stories.graphstory.com:7473/db/data/transaction node_labels: https://neo-55a53f8980d73-364459c455.do-stories.graphstory.com:7473/db/data/labels neo4j_version: 2.2.3) error nil
Test Case '-[TheoTests.Theo_000_RequestTests test_000_successfullyFetchDBMeta]' passed (0.465 seconds).
Test Suite 'Theo_000_RequestTests' passed at 2015-07-21 15:48:47 +0000.
	 Executed 1 test, with 0 failures (0 unexpected) in 0.465 (0.465) seconds
Test Suite 'TheoTests.xctest' passed at 2015-07-21 15:48:47 +0000.
	 Executed 1 test, with 0 failures (0 unexpected) in 0.465 (0.466) seconds
Test Suite 'Selected tests' passed at 2015-07-21 15:48:47 +0000.
	 Executed 1 test, with 0 failures (0 unexpected) in 0.465 (0.467) seconds
```

**Comment by nskostas** [comment]
The first mistake that I had, was that I didn't include "http://" in the Host url.
Then, it works for my localhost neo4j instance, but not for the graphstory/neo4j for some reason.

**Comment by jeremykendall** [comment]
@nskostas Make sure to double check your Graph Story Neo4j URL. The protocol should be `https://`.

**Comment by nskostas** [comment]
@jeremykendall Yes, you're right. Sorry, my mistake.  
Thank you! It works fine now!

**Comment by jeremykendall** [comment]
@nskostas No worries. I've done that a lot myself :-)

So glad to hear it's working! Nice job, @kwylez! Thanks for your excellent help troubleshooting, @nskostas!

Can we close this one out now?

**Comment by kwylez** [comment]
Ill merge this to master tonight.

**Comment by jbrozena22** [comment]
I'm getting this same issue when trying to access my test database on localhost. 

    `var theo: Client = Client(baseURL: "http://localhost:7474/") 
     theo.metaDescription({(meta, error) in
            print("meta in success \(meta) error \(error)")
     })`

RESULT:
`meta in success nil error Optional(Error Domain=com.theo.network.error Code=-1011 "(null)" UserInfo={TheoResponse=<NSHTTPURLResponse: 0x7fee826a8070> { URL: http://localhost:7474//db/data/ } { status code: 401, headers {
    "Content-Length" = 135;
    "Content-Type" = "application/json; charset=UTF-8";
    Date = "Tue, 01 Dec 2015 22:03:04 GMT";
    Server = "Jetty(9.2.z-SNAPSHOT)";
    "Www-Authenticate" = None;
} }, TheoResponseCode=401, NSLocalizedDescriptionKey=The response was empty, but you received at valid response code})
meta in success nil error Optional(Error Domain=com.theo.network.error Code=-1 "(null)" UserInfo={TheoResponse=<NSHTTPURLResponse: 0x7fee826a8070> { URL: http://localhost:7474//db/data/ } { status code: 401, headers {
    "Content-Length" = 135;
    "Content-Type" = "application/json; charset=UTF-8";
    Date = "Tue, 01 Dec 2015 22:03:04 GMT";
    Server = "Jetty(9.2.z-SNAPSHOT)";
    "Www-Authenticate" = None;
} }, TheoResponseCode=401, NSLocalizedDescriptionKey=There was an error processing the request})`

And obviously I can't use https with localhost. Any other suggestions? 

Xcode 7.1.1
Swift 2.1
Neo4j 2.3.1

**Comment by jeremykendall** [comment]
@jbrozena22 Are you sure you're passing the correct username and password? Have you visited http://localhost:7474/browser/ in your web browser to see if you have to change your username and password?

I ask because I frequently forget to change the default username and password when I develop locally.

**Comment by jbrozena22** [comment]
@jeremykendall I should have mentioned that I had tried it both ways - with and without authentication both with the same result. 

However I did discover the issue for anyone who runs into this same problem. The host URL needs to not end in a slash. So this worked:
`let theo: Client = Client(baseURL: "http://localhost:7474", user: "USERNAME", pass: "PASSWORD")`

But this didn't:
`let theo: Client = Client(baseURL: "http://localhost:7474/", user: "USERNAME", pass: "PASSWORD")`

(note the `/` at the end of localhost:7474)

Thanks!
